src/libAtomVM/nifs.c

```diff
@@ -2231,36 +2201,95 @@ static term nif_erlang_integer_to_binary_2(Context *ctx, int argc, term argv[])
         }
     }
 
-    avm_int64_t int_value = term_maybe_unbox_int64(value);
-    size_t len = lltoa(int_value, base, NULL);
+    _Static_assert(sizeof(intptr_t) >= sizeof(avm_int_t), "Cast to intptr_t is not safe");
 
-    if (UNLIKELY(memory_ensure_free_opt(ctx, term_binary_heap_size(len), MEMORY_CAN_SHRINK) != MEMORY_GC_OK)) {
+    if (term_is_integer(value)) {
+        avm_int_t int_val = term_to_int(value);
+        size_t wlen = intptr_write_to_ascii_buf(int_val, base, tmp_buf + tmp_buf_size);
+        *int_len = wlen;
+        *int_buf = tmp_buf + tmp_buf_size - wlen;
+    } else {
+        switch (term_boxed_size(value)) {
+            case 0:
+                UNREACHABLE();
+            case 1: {
+                avm_int_t int_val = term_unbox_int(value);
+                size_t wlen = intptr_write_to_ascii_buf(int_val, base, tmp_buf + tmp_buf_size);
+                *int_len = wlen;
+                *int_buf = tmp_buf + tmp_buf_size - wlen;
+                break;
+            }
+#if BOXED_TERMS_REQUIRED_FOR_INT64 == 2
+            case 2: {
+                avm_int64_t int64_val = term_unbox_int64(value);
+                size_t wlen = int64_write_to_ascii_buf(int64_val, base, tmp_buf + tmp_buf_size);
+                *int_len = wlen;
+                *int_buf = tmp_buf + tmp_buf_size - wlen;
+                break;
+            }
+#endif
+            default: {
+                size_t boxed_size = term_intn_size(value);
+                size_t digits_per_term = sizeof(term) / sizeof(intn_digit_t);
+                intn_digit_t *dest_buf = (void *) term_intn_data(value);
+                *int_buf = intn_to_string(dest_buf, boxed_size * digits_per_term, base, int_len);
+                *needs_cleanup = true;
+            }
+        }
+    }
+
+    return term_nil();
+}
+
+static term nif_erlang_integer_to_binary_2(Context *ctx, int argc, term argv[])
+{
+    size_t tmp_buf_size = INT64_WRITE_TO_ASCII_BUF_LEN;
+    char tmp_buf[tmp_buf_size];
+
+    char *int_buf;
+    size_t int_len;
+    bool needs_cleanup;
+    term maybe_fail_ret
+        = integer_to_buf(ctx, argc, argv, tmp_buf, tmp_buf_size, &int_buf, &int_len, &needs_cleanup);
+    if (UNLIKELY(term_is_invalid_term(maybe_fail_ret))) {
```

Comment:
Since maybe_fail_ret is not used, UNLIKELY could test the returned value of integer_to_buf directly.

Reply:
It is actually used:
```c
2254     if (UNLIKELY(term_is_invalid_term(maybe_fail_ret))) {
2255         return maybe_fail_ret;
2256     }
```

Reply:
Sorry for the confusion. What I mean is that we only test result being invalid term or not. When it's invalid term we return invalid term. In this respect we don't really use this value.

Reply:
oh, right, got it.